drivers/sensor/fdc2x1x/fdc2x1x.c

```diff
@@ -0,0 +1,1086 @@
+
+	return 0;
+}
+
+/**
+ * Initialization of the device.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int fdc2x1x_init(const struct device *dev)
+{
+	struct fdc2x1x_data *data = dev->data;
+	const struct fdc2x1x_config *cfg = dev->config;
+	uint8_t ch_supported;
+
+	if (cfg->fdc2x14) {
+		ch_supported = 4;
+	} else {
+		ch_supported = 2;
+	}
+
+	if (cfg->num_channels == 0) {
+		LOG_ERR("No channel nodes found");
+		return -EINVAL;
+	} else if (cfg->num_channels > ch_supported) {
+		LOG_ERR("Amount of channels not supported by this device");
+		return -EINVAL;
+	}
+
+	if (cfg->sd_name) {
+		if (fdc2x1x_init_sd_pin(dev) < 0) {
+			return -EINVAL;
+		}
+	}
+
+	data->bus = device_get_binding(cfg->i2c_name);
+	if (data->bus  == NULL) {
+		LOG_ERR("Failed to get pointer to %s device!", cfg->i2c_name);
+		return -EINVAL;
+	}
+
+	if (fdc2x1x_probe(dev) < 0) {
+		return -ENODEV;
+	}
+
+	if (fdc2x1x_reset(dev) < 0) {
+		return -EIO;
+	}
+
+	if (fdc2x1x_init_config(dev) < 0) {
+		return -EIO;
+	}
+
+	if (fdc2x1x_set_op_mode(dev, FDC2X1X_ACTIVE_MODE) < 0) {
+		return -EIO;
+	}
+#ifdef CONFIG_PM_DEVICE
+	data->pm_state = FDC2X1X_ACTIVE_MODE;
+#endif
+
+#ifdef CONFIG_FDC2X1X_TRIGGER
+	if (fdc2x1x_init_interrupt(dev) < 0) {
+		LOG_ERR("Failed to initialize interrupt!");
+		return -EIO;
+	}
+#endif
+
+	return 0;
+}
+
+#define FDC2X1X_SD_PROPS(n)			     \
+	.sd_name = DT_INST_GPIO_LABEL(n, sd_gpios),  \
+	.sd_pin = DT_INST_GPIO_PIN(n, sd_gpios),     \
+	.sd_flags = DT_INST_GPIO_FLAGS(n, sd_gpios), \
+
+#define FDC2X1X_SD(n)				       \
+	IF_ENABLED(DT_INST_NODE_HAS_PROP(n, sd_gpios), \
+		   (FDC2X1X_SD_PROPS(n)))
+
+#define FDC2X1X_INTB_PROPS(n)				 \
+	.intb_name = DT_INST_GPIO_LABEL(n, intb_gpios),	 \
+	.intb_pin = DT_INST_GPIO_PIN(n, intb_gpios),	 \
+	.intb_flags = DT_INST_GPIO_FLAGS(n, intb_gpios), \
+
+#define FDC2X1X_INTB(n)			   \
+	IF_ENABLED(CONFIG_FDC2X1X_TRIGGER, \
+		   (FDC2X1X_INTB_PROPS(n)))
+
+#define FDC2X1X_CH_CFG_INIT(ch)				   \
+	{						   \
+		.rcount = DT_PROP(ch, rcount),		   \
+		.offset = DT_PROP(ch, offset),		   \
+		.settle_count = DT_PROP(ch, settlecount),  \
+		.fref_divider = DT_PROP(ch, fref_divider), \
+		.idrive = DT_PROP(ch, idrive),		   \
+		.fin_sel = DT_PROP(ch, fin_sel),	   \
+		.inductance = DT_PROP(ch, inductance),	   \
+	},
+
+#define FDC2X1X_CHANNEL_BUF_INIT(ch)          0,
+
+#define FDC2X1X_INIT(n)							   \
+	static uint32_t fdc2x1x_sample_buf_##n[] = {			   \
+		DT_INST_FOREACH_CHILD(n, FDC2X1X_CHANNEL_BUF_INIT)	   \
+	};								   \
+									   \
+	static struct fdc2x1x_data fdc2x1x_data_##n = {			   \
+		.channel_buf = fdc2x1x_sample_buf_##n,			   \
+	};								   \
+									   \
+	const struct fdc2x1x_chx_config ch_cfg_##n[] = {		   \
+		DT_INST_FOREACH_CHILD(n, FDC2X1X_CH_CFG_INIT)		   \
+	};								   \
+									   \
+	static const struct fdc2x1x_config fdc2x1x_config_##n = {	   \
+		.i2c_name = DT_INST_BUS_LABEL(n),			   \
```

Comment:
Hi again. I just see one more round of churn before this PR LGTM from a driver model perspective.

We are making changes to the way that sensor devices access their buses for Zephyr 2.6. These mean that getting them by name using device_get_binding is discouraged.

You should now store the device directly in the configuration structure in ROM by moving these fields to `struct fdc2x1x_config` from `struct fdc2x1x_data`:

```
	const struct device *bus;
	const struct device *gpio;
```

These should now be initialized using `DEVICE_DT_GET`. For example, `.bus = DEVICE_DT_GET(DT_INST_BUS(n))`.

You should then check the device in the init routine using `device_is_ready`.

Finally, fields like `i2c_name` which exist only to store devicetree label properties should be removed. You can use `bus->name` instead if you need a name.

For an example, see the drivers/led/led_pwm.c changes in https://github.com/zephyrproject-rtos/zephyr/pull/32558.

Thanks for your patience!

Reply:
Hi. I addressed your comments. The bus and gpio devices are now initialized with `DEVICE_DT_GET` and checked  with `device_is_ready`. I also removed the `i2c_name` and both gpio labels.